sound/soc/sof/ipc.c

```diff
@@ -612,6 +623,169 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
+	cdata->cmd = ctrl_cmd;
+	cdata->type = ctrl_type;
+	cdata->rhdr.hdr.size = sipc->msg_bytes + sipc->hdr_bytes;
+	cdata->comp_id = scontrol->comp_id;
+	cdata->msg_id = 0;
+	cdata->num_elems = cdata->total_elems;
+
+	return 0;
+}
+
+static int sof_get_ctrl_copy_params(enum sof_ipc_ctrl_type ctrl_type,
+				    struct sof_ipc_ctrl_data *cdata,
+				    struct sof_ipc_ctrl_data *hdr,
+				    struct sof_ctrl_data_send_params *sipc)
+{
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sipc->src = (unsigned char *)cdata->chanv;
+		sipc->dst = (unsigned char *)hdr->chanv;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sipc->src = (unsigned char *)cdata->compv;
+		sipc->dst = (unsigned char *)hdr->compv;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sipc->src = (unsigned char *)cdata->data->data;
+		sipc->dst = (unsigned char *)hdr->data->data;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* calculate payload size and number of messages */
+	sipc->pl_size = SOF_IPC_MSG_MAX_SIZE - sipc->hdr_bytes;
+	sipc->num_msg = DIV_ROUND_UP(sipc->msg_bytes, sipc->pl_size);
+	sipc->offset = 0;
+
+	return 0;
+}
+
+static int sof_send_recv_ctrl_data(struct snd_sof_dev *sdev,
+				   struct snd_sof_control *scontrol,
+				   u32 ipc_cmd,
+				   enum sof_ipc_ctrl_type ctrl_type,
+				   enum sof_ipc_ctrl_cmd ctrl_cmd)
+{
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct sof_ipc_ctrl_data *hdr;
+	struct sof_ctrl_data_send_params sipc;
+	int send;
+	int err = 0;
+	int i;
+
+	/* calculate header and data size */
+	sof_get_ctrl_size_params(scontrol, ipc_cmd, ctrl_type, ctrl_cmd, cdata,
+				 &sipc);
+
+	/* send normal size ipc */
+	if (cdata->rhdr.hdr.size <= SOF_IPC_MSG_MAX_SIZE)
+		return sof_ipc_tx_message(sdev->ipc, cdata->rhdr.hdr.cmd, cdata,
+					  cdata->rhdr.hdr.size, cdata,
+					  cdata->rhdr.hdr.size);
+
+	/* if bigger than max ipc size, chop to smaller pieces */
+	dev_dbg(sdev->dev, "large ipc size %u, control size %u\n",
+		cdata->rhdr.hdr.size, scontrol->size);
+
+	/* allocate max ipc size because we have at least one */
+	hdr = (struct sof_ipc_ctrl_data *)
+		kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!hdr)
+		return -ENOMEM;
+
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_DATA_GET:
+		sof_get_ctrl_copy_params(ctrl_type, hdr, cdata, &sipc);
+		send = 0;
+		break;
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sof_get_ctrl_copy_params(ctrl_type, cdata, hdr, &sipc);
+		send = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* copy the header data */
+	memcpy(hdr, cdata, sipc.hdr_bytes);
+
+	/* copy the data in a loop */
+	for (i = 0; i < sipc.num_msg; i++) {
+		hdr->msg_id = i;
+		sipc.send_bytes = sipc.msg_bytes > sipc.pl_size ? sipc.pl_size :
```

Comment:
@juimonen why does send_bytes need to be a member of this struct? Cant it just be a local variable here?

Reply:
Also do I understand correctly that the last part of the message will always be of pl_size even if it is actually smaller than that?

Reply:
should not... here's trace from sending large blob to fir:
ipc: hdr 0x50030000 tx (384) > rx (92)
ipc: comp 3 -> cmd 3
eq_fir_cmd()
fir_cmd_set_data(), SOF_CTRL_CMD_BINARY
fir_cmd_set_data(): blob size: 260 msg_id 0

ipc: hdr 0x50030000 tx (384) > rx (92)
ipc: comp 3 -> cmd 3
eq_fir_cmd()
fir_cmd_set_data(), SOF_CTRL_CMD_BINARY
fir_cmd_set_data(): blob size: 260 msg_id 1

ipc: hdr 0x50030000 tx (384) > rx (92)
ipc: comp 3 -> cmd 3
eq_fir_cmd()
fir_cmd_set_data(), SOF_CTRL_CMD_BINARY
fir_cmd_set_data(): blob size: 260 msg_id 2

ipc: hdr 0x50030000 tx (384) > rx (92)
ipc: comp 3 -> cmd 3
eq_fir_cmd()
fir_cmd_set_data(), SOF_CTRL_CMD_BINARY
fir_cmd_set_data(): blob size: 260 msg_id 3

ipc: hdr 0x50030000 tx (156) > rx (92)
ipc: comp 3 -> cmd 3
eq_fir_cmd()
fir_cmd_set_data(), SOF_CTRL_CMD_BINARY
fir_cmd_set_data(): blob size: 32 msg_id 4